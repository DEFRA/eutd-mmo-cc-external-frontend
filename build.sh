#/bin/bash
curl -O https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz
tar -xf node-v12.18.4-linux-x64.tar.xz --directory /tmp
export PATH="/tmp/node-v12.18.4-linux-x64/bin:$PATH"
rm node-v12.18.4-linux-x64.tar.xz
npm ci
npm run test-vstack
rm -rf /tmp/node-v12.18.4-linux-x64