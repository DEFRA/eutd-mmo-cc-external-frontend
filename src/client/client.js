// Startup point for the client side application
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Router, withRouter} from 'react-router-dom';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import axios from 'axios';
import Routes from './Routes';
import reducers from './reducers';
import { createBrowserHistory } from 'history';

const composeEnhancers = process.env.NODE_ENV === 'development' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__  ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose;
  

const paramSerializerFn = params => {
  let output = '';
  Object.keys(params).forEach((k) => {
    output+= `${k}=${params[k]}&`;
  });
  return encodeURI(output);
};

const axiosResponseInterceptorSuccessFn = response => response;
const axiosResponseInterceptorErrorFn = error => {
  if (error.response.status === 599) {
    // Let's check if this is because we're logged out
    const empPromise = new Promise(() => {});
    const searchParam = error.request.responseURL.slice(error.request.responseURL.search(/[?]/g));
    let searchParams = new URLSearchParams(searchParam);
    if(searchParams.get('notLoggedInErr') === 'yes') {
      window.location = '/logout?backToPath=/server-logout';
      return empPromise;
    }
    // It's always a timeout...
    window.location = '/timed-out';
    return new Promise(() => {});
  } else {
    return new Promise((_,rej) => rej(error));
  }
};

const axiosInstanceForOrchService = axios.create({
  baseURL: '/orchestration/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForOrchService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForRefService = axios.create({
  baseURL: '/reference/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForRefService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForDynamixService = axios.create({
  baseURL: '/dynamix/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForDynamixService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const store = createStore(
  reducers,
  window.INITIAL_STATE,
  composeEnhancers(applyMiddleware(
    // The difference between orch/ref service api is not significant as they both go through proxy
    // However it is significant in the server side store. To stick with same action creators we're
    // passing down same axios instance to both orchestration and reference service api .
    thunk.withExtraArgument( {
      orchestrationApi: axiosInstanceForOrchService,
      referenceServiceApi: axiosInstanceForRefService,
      dynamixApi: axiosInstanceForDynamixService } )
  ))
);

const ScrollToTop = withRouter( class ScrollToTop extends React.Component {
  componentDidMount() {
  }

  render() {
    return this.props.children;
  }
});

const history = createBrowserHistory();


ReactDOM.hydrate(
  <Provider store={store}>
    <Router history={history}>
      <ScrollToTop>
        <div>{renderRoutes(Routes)}</div>
      </ScrollToTop>
    </Router>
  </Provider>,
  document.querySelector('#root')
);
