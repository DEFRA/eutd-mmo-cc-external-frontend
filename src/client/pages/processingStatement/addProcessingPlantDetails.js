import React from 'react';
import { connect } from 'react-redux';
import _ from 'lodash';

import {
  Main,
  BackLink,
  InputField,
  Header,
  GridRow,
  GridCol
} from 'govuk-react';

import {
  getProcessingStatementFromRedis,
  saveProcessingStatement,
  saveProcessingStatementToRedis,
  clearProcessingStatement
} from '../../actions';

import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import { withTranslation } from 'react-i18next';

import {
  toGovukErrors,
  onHandleErrorClick,
  scrollToErrorIsland
} from '../utils';
import ErrorIsland from '../../components/elements/ErrorIsland';
import PageTitle from '../../components/PageTitle';
import Form from '../../components/elements/Form';
import {withRouter} from 'react-router-dom';
import HelpLink from '../../components/HelpLink';
import SaveAsDraftButton from '../../components/SaveAsDraftButton';
import ContinueButton from '../../components/elements/ContinueButton';
import BackToProgressLink from '../../components/BackToProgressLink';

class AddProcessingPlantDetails extends React.Component {
  async save() {
    const {route, match} = this.props;
    const documentNumber = match.params.documentNumber;
    const processingStatement = await this.props.saveToRedis(
      {
        data: this.props.processingStatement || {},
        currentUrl: route.path,
        saveAsDraft: false,
        documentNumber: documentNumber
      });

    if (!_.isEmpty(processingStatement.errors)) return scrollToErrorIsland();
    this.props.history.push(route.nextUri.replace(':documentNumber', documentNumber));
  }

  onChange(e) {
    const processingStatement = _.cloneDeep(this.props.processingStatement || {});
    processingStatement[e.target.name] = e.target.value;
    this.props.save(processingStatement);
  }

  goBack(e, page) {
    e.preventDefault();
    this.props.history.push(page);
  }

  onSaveAsDraft = async (e) => {
    e.preventDefault();
    const { route } = this.props;
    const { documentNumber } = this.props.match.params;
    const { path, saveAsDraftUri } = route;

    await this.props.saveToRedis(
      {
        data: this.props.processingStatement || {},
        currentUrl: path,
        saveAsDraft: true,
        documentNumber: documentNumber
      });

    this.props.history.push(saveAsDraftUri);
  }

  redirectToForbidden() {
    this.props.history.push('/forbidden');
  }

  componentWillUnmount() {
    this.props.clear();
  }

  async componentDidMount() {
    await this.props.getFromRedis(this.props.match.params.documentNumber);
  }

  componentDidUpdate() {
    const { processingStatement } = this.props;
    if (processingStatement.unauthorised === true) {
      this.redirectToForbidden();
    }
  }

  formatErrorResponse(error) {
    let errorObject = {};
    if(!_.isEmpty(error)) {
      Object.keys(error).forEach(key => {
        return errorObject[key] = this.errorFormatForPersonResponsible(error[key]);
      });
    }
    return errorObject;
  }

  errorFormatForPersonResponsible(error) {
    const {t} = this.props;
    if (!_.isEmpty(error)) {
      const splitDynamicErrors = error.split('-');
      if (splitDynamicErrors.length > 0) {
        return t(splitDynamicErrors[0], {
          MAX_PERSON_RESPONSIBLE_LENGTH: splitDynamicErrors[1],
        });
      }
    }
    return t(error);
  }

  render() {
    const { processingStatement = {}, history, route, match, t } = this.props;
    const { previousUri, path, nextUri, saveAsDraftUri, journey, progressUri, title } = route;
    const { documentNumber } = match.params;
    const errors = (processingStatement.errorsUrl === history.location.pathname && processingStatement.errors) ? processingStatement.errors : {}; // only display errors for this URL
    const { personResponsibleForConsignment = '', plantApprovalNumber = '' } = processingStatement;
    const saveAsDraftFormActionUrl = `/orchestration/api/v1/processingStatement/saveAndValidate?c=${path}&n=${nextUri}&saveAsDraftUrl=${saveAsDraftUri}`;
    const backUrl = previousUri.replace(':documentNumber', documentNumber);

    return (
      <Main className="processing-statement">
        <PageTitle title={`${!_.isEmpty(errors) ? 'Error: ' : ''}${t('psAddProcessingPDAddProcessingPlantDetails')} - ${t(title)}`} />
        <ErrorIsland errors={toGovukErrors(this.formatErrorResponse(errors))} onHandleErrorClick={onHandleErrorClick}/>
        <GridRow>
          <GridCol>
            <BackLink onClick={e => this.goBack(e, backUrl)} href={backUrl}>
            {t('commonBackLinkBackButtonLabel')}
            </BackLink>
            <Header>{t('psAddProcessingPDAddProcessingPlantDetails')}</Header>
          </GridCol>
        </GridRow>
        <Form
          action="/orchestration/api/v1/processingStatement/saveAndValidate"
          currentUrl={path}
          nextUrl={nextUri}
          onSubmit={e => e.preventDefault()}
        >
          <GridRow>
            <GridCol columnTwoThirds>
              <InputField
                id="personResponsibleForConsignment"
                meta={{error: this.errorFormatForPersonResponsible(errors['personResponsibleForConsignment']), touched: true}}
                htmlFor={'personResponsibleForConsignment'}
                input={{
                  autoComplete: 'off',
                  className: 'formControl',
                  name: 'personResponsibleForConsignment',
                  id: 'personResponsibleForConsignment',
                  value: personResponsibleForConsignment,
                  onChange: e => this.onChange(e, 'personResponsibleForConsignment')
                }}>{t('psAddProcessingPDPersonResponsibleForThisConsignment')}</InputField>
            </GridCol>
          </GridRow>

          <GridRow>
            <GridCol columnTwoThirds>
              <InputField
                hint={t('psAddProcessingPDHintTextForPlantApprovalNumber')}
                id="plantApprovalNumber"
                meta={{error: t(errors['plantApprovalNumber']), touched: true}}
                htmlFor={'plantApprovalNumber'}
                input={{
                  autoComplete: 'off',
                  className: 'formControl',
                  name: 'plantApprovalNumber',
                  id: 'plantApprovalNumber',
                  value: plantApprovalNumber,
                  onChange: e => this.onChange(e, 'plantApprovalNumber')
                }}>{t('psAddProcessingPDPlantApprovalNumber')}</InputField>
            </GridCol>
          </GridRow>
          <GridRow>
            <SaveAsDraftButton formactionUrl={saveAsDraftFormActionUrl} onClick={this.onSaveAsDraft} />
            <input type="hidden" name="journey" value={journey} />
            <input type="hidden" name="currentUri" value={path} />
            <ContinueButton id="continue" onClick={() => this.save()}>{t('commonContinueButtonSaveAndContinueButton')}</ContinueButton>
          </GridRow>
        </Form>
        <BackToProgressLink
          progressUri={progressUri}
          documentNumber={documentNumber}
        />
        <HelpLink journey={journey}/>
      </Main>
      );
    }

}


function mapStateToProps(state) {
  return {
    processingStatement: state.processingStatement
  };
}

function loadData(store) {
  return store.dispatch(getProcessingStatementFromRedis(this.documentNumber));
}

export const component = withRouter(
  connect(mapStateToProps, {
    save: saveProcessingStatement,
    clear: clearProcessingStatement,
    saveToRedis: saveProcessingStatementToRedis,
    getFromRedis: getProcessingStatementFromRedis
  })(withTranslation() (AddProcessingPlantDetails)) );

export default {
  loadData,
  component: PageTemplateWrapper( component )
};
