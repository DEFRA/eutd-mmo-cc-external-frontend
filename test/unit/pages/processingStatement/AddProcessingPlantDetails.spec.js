import * as React from 'react';
import { mount } from 'enzyme/build';
import { Provider } from 'react-redux';
import { Router, Route } from 'react-router-dom';
import { act } from 'react-dom/test-utils';
import { createMemoryHistory } from 'history';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import AddProcessingPlantDetailsWrapper from '../../../../src/client/pages/processingStatement/addProcessingPlantDetails';
import * as ErrorUtils from '../../../../src/client/pages/utils/errorUtils';
import * as clientAction from '../../../../src/client/actions/index.js';
import { getProcessingStatementFromRedis, clearProcessingStatement } from '../../../../src/client/actions';
import { component as AddProcessingPlantDetails } from '../../../../src/client/pages/processingStatement/addProcessingPlantDetails';
import { render } from '@testing-library/react';

jest.mock('../../../../src/client/actions');


describe('Add processing plant details page', () => {
  const mockStore = configureStore([thunk]);
  const mockClearErrors = jest.fn();

  getProcessingStatementFromRedis.mockReturnValue({
    type: 'GET_PROCESSING_STATEMENT_FROM_REDIS',
  });

  clearProcessingStatement.mockReturnValue({
    type: 'CLEAR_PROCESSING_STATEMENT',
  });

  let wrapper;
  it('should redirect to the previous page when required data is missing', async () => {
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment: 'psAddProcessingPlantDetailsErrorNullReponsiblePerson',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber'
        },
      },
      global: {
        allFish: []
      },
      t: jest.fn()
    });


    const props = {
      clear: mockClearErrors,
      t: jest.fn()
    };

    await act(async () => {
      wrapper = await mount(
        <Provider store={store}>
          <Router history={history}>
            <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
              <AddProcessingPlantDetails
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                  progressUri: '/create-processing-statement/:documentNumber/progress',
                  path: 'path',
                  nextUri: 'nextUri'
                }}
                {...props}
              />
            </Route>
          </Router>
        </Provider>
      );
    });
    expect(mockPush).not.toHaveBeenCalled();
  });

  it('should generate snapshot', ()=> {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  });

});

describe('English/Welsh Translation test cases', () => {
  const mockStore = configureStore([thunk]);
  const mockClearErrors = jest.fn();
  const documentNumber = 'document123';
  const mockSaveToRedis = jest.spyOn(clientAction, 'saveProcessingStatementToRedis');
  const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });

  const mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');
  mockScrollToErrorIsland.mockReturnValue(null);
  let wrapper;
  beforeEach(() => {
    mockClearErrors.mockReset();
    mockScrollToErrorIsland.mockReset();
    mockScrollToErrorIsland.mockImplementation(() => null);

    const store = mockStore({
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment: 'psAddProcessingPlantDetailsErrorNullReponsiblePerson',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber'
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-processing-plant-details`
      },
      global: {
        allFish: []
      },
      t: jest.fn()
    });

  const props = {
    clear: mockClearErrors,
    t: jest.fn()
  };

  wrapper =  mount(
    <Provider store={store}>
      <Router history={history}>
        <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
          <AddProcessingPlantDetails
            route={{
              previousUri:
                '/create-processing-statement/:documentNumber/previous-uri',
              progressUri: '/create-processing-statement/:documentNumber/progress',
              path: 'path',
              nextUri: 'nextUri'
            }}
            {...props}
          />
        </Route>
      </Router>
    </Provider>
   );
  });

  it('when unauthorised is false', async () => {
    let  wrapper;
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: {
        unauthorised: false
      }
    });

    wrapper = await mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
              <AddProcessingPlantDetails
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                  progressUri: '/create-processing-statement/:documentNumber/progress',
                }}
              />
          </Route>
        </Router>
      </Provider>
    );

    await act(async () => await wrapper);
    expect(wrapper).toBeDefined();
  });

  it('when processingStatement is undefined', async () => {
    let  wrapper;
    const documentNumber = 'document123';
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);

    const store = mockStore({
      processingStatement: undefined
    });

    wrapper = await mount(
      <Provider store={store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
              <AddProcessingPlantDetails
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                  progressUri: '/create-processing-statement/:documentNumber/progress',
                }}
              />
          </Route>
        </Router>
      </Provider>
    );
    await act(async () => await wrapper);
    expect(wrapper).toBeDefined();
  });

  it('should be scroll to errorIsLand', async () => {
    const documentNumber = 'document123';
    let mockScrollToErrorIsland;
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });
    mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');
    const mockPush = jest.spyOn(history, 'push');
    mockPush.mockReturnValue(null);
    mockSaveToRedis.mockReturnValue({
      errors: {
        personResponsibleForConsignment: 'psAddProcessingPlantDetailsErrorNullReponsiblePerson-50',
        plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber'
      },
    });

    const props = {
      route: {
        path: ':documentNumber/add-processing-plant-details',
        previousUri: ':documentNumber/catch-added',
        progressUri: '/create-processing-statement/:documentNumber/progress',
        nextUri: ':documentNumber/add-processing-plant-address',
        journey:  'processingStatement',
        title: 'Create a UK processing statement - GOV.UK',
        saveAsDraftUri: 'processing-statements'
      },
      match:{
        params: {documentNumber}
      },
      processingStatement: undefined,
      history: {
        ...history,
        location: {
          pathname: `/create-processing-statement/${documentNumber}/add-processing-plant-details`
        }
      },
      clear: mockClearErrors,
      t: jest.fn(),
      saveToRedis: mockSaveToRedis
    };

    await new AddProcessingPlantDetails.WrappedComponent({...props}).save('documentNumber/add-processing-plant-address');

    expect(mockSaveToRedis).toHaveBeenCalledWith({
      data: props.processingStatement || {},
      currentUrl: ':documentNumber/add-processing-plant-details',
      saveAsDraft: false,
      documentNumber: documentNumber
    });

    expect(mockScrollToErrorIsland).toHaveBeenCalled();
  });

  it('will call all methods needed to load the component', async () => {
    const store = {
      dispatch: () => {
        return new Promise((resolve) => {
          resolve();
        });
      }
    };

    await AddProcessingPlantDetailsWrapper.loadData(store, 'processingStatement');
    expect(getProcessingStatementFromRedis).toHaveBeenCalled();
  });

  it('will call the the on submit for component', () => {
  const event = {
      preventDefault: jest.fn()
    };
    const formEl = wrapper.find('form');
    formEl.props().onSubmit(event);

    expect(formEl.exists()).toBeTruthy();
    expect(event.preventDefault).toHaveBeenCalled();
 });


 it('should clear errors on component will unmount', () => {
  const mockDispatchClearErrors = jest.fn();

  new AddProcessingPlantDetails.WrappedComponent({
    match: {
      params: {
        documentNumber: ''
      }
    },
    unauthorised: true,
    clear: mockDispatchClearErrors
  }).componentWillUnmount();

  expect(mockDispatchClearErrors).toHaveBeenCalled();
});

it('should call on onChange/goBack when entering a description', async () => {
  const documentNumber = 'document123';
  const history = createMemoryHistory({
    initialEntries: [
      `/create-processing-statement/${documentNumber}/add-processing-plant`,
    ],
  });
  const mockPush = jest.spyOn(history, 'push');
  mockPush.mockReturnValue(null);
  mockSaveToRedis.mockReturnValue({
    errors: {
      personResponsibleForConsignment: 'psAddProcessingPlantDetailsErrorNullReponsiblePerson',
      plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber'
    },
  });

  const props = {
    route: {
      path: ':documentNumber/add-processing-plant-details',
      previousUri: ':documentNumber/catch-added',
      nextUri: ':documentNumber/add-processing-plant-address',
      journey:  'processingStatement',
      title: 'Create a UK processing statement - GOV.UK',
      saveAsDraftUri: 'processing-statements'
    },
    match:{
      params: {documentNumber}
    },
    processingStatement: undefined,
    history: {
      ...history,
      location: {
        pathname: `/create-processing-statement/${documentNumber}/add-processing-plant`
      }
    },
    clear: mockClearErrors,
    t: jest.fn(),
    saveToRedis: mockSaveToRedis,
    save: jest.fn()
  };

  await new AddProcessingPlantDetails.WrappedComponent({...props}).onChange({
    target:{
      name: 'consignmentDescription'
    }
  });

  await new AddProcessingPlantDetails.WrappedComponent({...props}).goBack({
    preventDefault: jest.fn()
  });

  expect(props.save).toHaveBeenCalled();
});

});

describe('Proccessing plant details errors', () => {
  const mockStore = configureStore([thunk]);
  const documentNumber = 'document123';
  let wrapper;

  it('Should show error section if condition met', async() => {
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });
    const store = mockStore({
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment:
            'psAddProcessingPlantDetailsErrorNullReponsiblePerson',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber',
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-processing-plant`,
      },
      global: {
        allFish: [],
      },
      t: jest.fn(),
    });

    const props = {
      history:{
        ...history,
        location:{
          pathname: `/create-processing-statement/${documentNumber}/add-processing-plant`
        }
      },
      clear: jest.fn(),
      t: jest.fn(),
    };

    await act(async () => {
      wrapper = await mount(
        <Provider store={store}>
          <Router history={history}>
            <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
              <AddProcessingPlantDetails
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                  progressUri: '/create-processing-statement/:documentNumber/progress',
                  path: 'path',
                  nextUri: 'nextUri'
                }}
                {...props}
              />
            </Route>
          </Router>
        </Provider>
      );
    });

    expect(wrapper.find('ErrorIsland')).toBeDefined();
  });

  it('Should check dynamic error if condition met', async() => {
    const history = createMemoryHistory({
      initialEntries: [
        `/create-processing-statement/${documentNumber}/add-processing-plant`,
      ],
    });
    const store = mockStore({
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment:
            'psAddProcessingPDErrorPersonResponsibleForConsignmentLength-50',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber',
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-processing-plant`,
      },
      global: {
        allFish: [],
      },
      t: jest.fn(),
      errorFormatForPersonResponsible: jest.fn()
    });

    const props = {
      history:{
        ...history,
        location:{
          pathname: `/create-processing-statement/${documentNumber}/add-processing-plant`
        }
      },
      clear: jest.fn(),
      t: jest.fn(),
    };

    await act(async () => {
      wrapper = await mount(
        <Provider store={store}>
          <Router history={history}>
            <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
              <AddProcessingPlantDetails
                route={{
                  previousUri:
                    '/create-processing-statement/:documentNumber/previous-uri',
                  progressUri: '/create-processing-statement/:documentNumber/progress',
                  path: 'path',
                  nextUri: 'nextUri'
                }}
                {...props}
              />
            </Route>
          </Router>
        </Provider>
      );
    });
    expect(wrapper.find('ul.error-summary-list li a').at(0).text()).toBe('Person responsible for this consignment must not exceed 50 characters');
  });
});

describe('Errors after translation', () => {
  const mockStore = configureStore([thunk]);
  const mockScrollToErrorIsland = jest.spyOn(ErrorUtils, 'scrollToErrorIsland');
  mockScrollToErrorIsland.mockReturnValue(null);

  getProcessingStatementFromRedis.mockReturnValue({
    type: 'GET_PROCESSING_STATEMENT_FROM_REDIS',
  });

  const documentNumber = 'document123';
  const history = createMemoryHistory({
    initialEntries: [
      `/create-processing-statement/${documentNumber}/add-processing-plant`,
    ],
  });
  const mockPush = jest.spyOn(history, 'push');
  mockPush.mockReturnValue(null);

  const getComponentWrapper = (store={}, props={}) => {
    const defaultStore = {
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment: 'psAddProcessingPDErrorPersonResponsibleForConsignment',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber',
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-processing-plant`,
        personResponsibleForConsignment: '',
        plantApprovalNumber: ''
      },
      global: {
        allFish: [],
      },
      t: jest.fn(),
      ...store
    };

    const _store = mockStore(defaultStore);

    const _props = {
      history:{
        ...history,
        location:{
          pathname: `/create-processing-statement/${documentNumber}/add-processing-plant`
        }
      },
      clear: jest.fn(),
      t: jest.fn(),
      ...props
    }

    return mount(
      <Provider store={_store}>
        <Router history={history}>
          <Route path="/create-processing-statement/:documentNumber/add-processing-plant">
            <AddProcessingPlantDetails
              route={{
                previousUri:
                  '/create-processing-statement/:documentNumber/previous-uri',
                progressUri: '/create-processing-statement/:documentNumber/progress',
                path: 'path',
                nextUri: 'nextUri'
              }}
              {..._props}
            />
          </Route>
        </Router>
      </Provider>
    );
  }

  it('should be scroll to errorIsLand with empty form submission', () => {
    const wrapper = getComponentWrapper();
    const event = {
      preventDefault: jest.fn(),
    };
    const formEl = wrapper.find('form');
    formEl.props().onSubmit(event);

    expect(wrapper.find('ul.error-summary-list li a').at(0).text()).toBe('Enter the name of the person responsible for this consignment');
    expect(wrapper.find('ul.error-summary-list li a').at(1).text()).toBe('Enter the plant approval number');
  });

  it('should be scroll to errorIsLand with false value submission', () => {
    const store = {
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment: 'psAddProcessingPDErrorResponsibleValidation',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber',
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-processing-plant`,
        personResponsibleForConsignment: '<!2##$!>'
      },
    };
    const wrapper = getComponentWrapper(store);
    const event = {
      preventDefault: jest.fn(),
    };
    const formEl = wrapper.find('form');
    formEl.props().onSubmit(event);

    expect(wrapper.find('ul.error-summary-list li a').at(0).text()).toBe('Person responsible must only contain letters, hyphens, spaces, and single quotes');
  });
  it('should be scroll to errorIsLand with dynamic error', () => {
    const store = {
      processingStatement: {
        catches: [],
        consignmentDescription: 'consignment',
        errors: {
          personResponsibleForConsignment: 'psAddProcessingPDErrorPersonResponsibleForConsignmentLength-50',
          plantApprovalNumber: 'psAddProcessingPDErrorPlantApprovalNumber',
        },
        errorsUrl: `/create-processing-statement/${documentNumber}/add-processing-plant`,
        personResponsibleForConsignment: '<!2##$!>'
      },
    };
    const wrapper = getComponentWrapper(store);
    const event = {
      preventDefault: jest.fn(),
    };
    const formEl = wrapper.find('form');
    formEl.props().onSubmit(event);
    expect(wrapper.find('AddProcessingPlantDetails').instance().errorFormatForPersonResponsible(store.processingStatement.errors['personResponsibleForConsignment'])).toBe('Person responsible for this consignment must not exceed 50 characters');
  });

});