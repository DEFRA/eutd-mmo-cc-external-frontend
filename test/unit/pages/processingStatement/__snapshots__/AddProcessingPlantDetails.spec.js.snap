// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add processing plant details page should generate snapshot 1`] = `
<div>
  <div
    class="processing-statement css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="/create-processing-statement/document123/previous-uri"
          >
            Back
          </a>
          <h1
            class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
          >
            Add processing plant details
          </h1>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/processingStatement/saveAndValidate?c=path&n=nextUri"
        method="POST"
      >
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <label
              class="e1vcssft0 e1vcssft0 css-5tg97n-StyledLabel-StyledHoc-StyledHoc e4u1u2g0"
              for="personResponsibleForConsignment"
              id="personResponsibleForConsignment"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Person responsible for this consignment
              </span>
              
              <input
                autocomplete="off"
                class="formControl e1vcssft0 css-4n7f6s-StyledInput-StyledHoc e4u1u2g0"
                id="personResponsibleForConsignment"
                name="personResponsibleForConsignment"
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <label
              class="e1vcssft0 e1vcssft0 css-5tg97n-StyledLabel-StyledHoc-StyledHoc e4u1u2g0"
              for="plantApprovalNumber"
              id="plantApprovalNumber"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              >
                Plant approval number
              </span>
              <span
                class="e1vcssft0 css-1whl0n4-StyledHint-StyledHoc e4u1u2g0"
              >
                This is sometimes called a site code.
              </span>
              
              <input
                autocomplete="off"
                class="formControl e1vcssft0 css-4n7f6s-StyledInput-StyledHoc e4u1u2g0"
                id="plantApprovalNumber"
                name="plantApprovalNumber"
                type="text"
                value=""
              />
            </label>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <button
            class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
            formaction="/orchestration/api/v1/processingStatement/saveAndValidate?c=path&n=nextUri&saveAsDraftUrl=undefined"
            id="saveAsDraft"
            name="saveAsDraft"
            type="submit"
          >
            Save as Draft
          </button>
          <input
            name="journey"
            type="hidden"
            value=""
          />
          <input
            name="currentUri"
            type="hidden"
            value="path"
          />
          <button
            class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
            id="continue"
          >
            Save and continue
          </button>
        </div>
      </form>
      <a
        href="/create-processing-statement/document123/progress"
        id="backToProgress"
      >
        Back to Your Progress
        <span
          class="govuk-visually-hidden"
        >
          back to your progress page link
        </span>
      </a>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
